dotnet-engine/UnleashEngine/UnleashEngine.cs

```diff
@@ -5,99 +5,142 @@
     {
-        engine_free(this.state);
+        platformEngine.FreeEngine(this.state);
         GC.SuppressFinalize(this);
     }
 
     public void TakeState(string json)
     {
-        engine_take_state(state, json);
+        var takeStatePtr = platformEngine.TakeState(state, json);
+
+        if (takeStatePtr == IntPtr.Zero)
+        {
+            return;
+        }
+
+        var takeStateJson = Marshal.PtrToStringUTF8(takeStatePtr);
+
+        platformEngine.FreeResponse(takeStatePtr);
+
+        var takeStateResult = takeStateJson != null ?
+            JsonSerializer.Deserialize<EngineResponse>(takeStateJson, options) :
+            null;
+
+        if (takeStateResult?.StatusCode == "Error") {
+            throw new UnleashException($"Error: {takeStateResult?.ErrorMessage}");
+        }
     }
 
     public bool IsEnabled(string toggleName, Context context)
     {
         string contextJson = JsonSerializer.Serialize(context, options);
-        return engine_is_enabled(state, toggleName, contextJson);
+        var isEnabledPtr = platformEngine.CheckEnabled(state, toggleName, contextJson);
+
+        if (isEnabledPtr == IntPtr.Zero)
+        {
+            return false;
+        }
+
+        var isEnabledJson = Marshal.PtrToStringUTF8(isEnabledPtr);
+
+        platformEngine.FreeResponse(isEnabledPtr);
+
+        var isEnabledResult = isEnabledJson != null ?
+            JsonSerializer.Deserialize<EngineResponse<bool?>>(isEnabledJson, options) :
+            null;
+
+
+        if (isEnabledResult?.StatusCode == "Error") {
+            throw new UnleashException($"Error: {isEnabledResult?.ErrorMessage}");
+        }
+
+        var enabled = isEnabledResult?.Value ?? false;
+
+        platformEngine.CountToggle(state, toggleName, enabled);
+
+        return enabled;
     }
 
     public Variant? GetVariant(string toggleName, Context context)
     {
         var contextJson = JsonSerializer.Serialize(context, options);
-        var variantPtr = engine_get_variant(state, toggleName, contextJson);
+        var variantPtr = platformEngine.CheckVariant(state, toggleName, contextJson);
 
         if (variantPtr == IntPtr.Zero)
         {
             return null;
         }
-        else
-        {
-            var variantJson = Marshal.PtrToStringUTF8(variantPtr);
 
-            var variant = JsonSerializer.Deserialize<Variant>(variantJson, options);
+        var variantJson = Marshal.PtrToStringUTF8(variantPtr);
 
-            engine_free_variant_def(variantPtr);
+        platformEngine.FreeResponse(variantPtr);
 
-            return variant;
+        var variantResult = variantJson != null ?
+            JsonSerializer.Deserialize<EngineResponse<Variant>>(variantJson, options) :
+            null;
+
+        if (variantResult?.StatusCode == "Error") {
+            throw new UnleashException($"Error: {variantResult?.ErrorMessage}");
         }
+
+        var variant = variantResult?.Value ?? Variant.DISABLED_VARIANT;
```

Comment:
Here we should not return the disabled variant because the user can provide a default value when the feature is not found. So it should be an SDK responsibility. We should either return null or an error. Most likely a null

Reply:
Ooo good spot!

Reply:
Actually, I'm thinking about this again, and if UnleashEngine does not behave as an SDK and we're saying that the SDK should take care of some details, then how can we test it against the specification?

Reply:
> then how can we test it against the specification

Yeah this is what caused me to return a default disabled variant, spec was failing.

I wonder if we on the SDK-side should check if it's the disabled variant being returned and if so and a default has been requested return the default. But I'm open to suggestions